internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,118 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) string
```

Comment:
```suggestion
	Token(context.Context) (string, error)
```

And handle errors correctly, return them via this method.